designs/s3/20200210-design-vscode-s3.md

```diff
@@ -0,0 +1,146 @@
+
+Each region of the Toolkit explorer tree includes an S3 node (see mockup
+below). The toolkit pulls a list of names and properties of S3 objects in the
+currently active profile and presents this data as a tree in the toolkit
+explorer. This is typical of existing toolkit features such as "Schemas" and
+"CDK".
+
+- When a region is expanded, make an eager request to get the count of S3
+  buckets for that region.
+- Show the count of buckets next to the S3 node.
+- The S3 node is always present. If there are no buckets in the region, "grey
+  out" the S3 node.
+- When the current active profile changes, the tree is not refreshed, but is
+  greyed-out to indicate that it is stale. It is not minimized.
+- When a bucket or object is created or deleted, only that node is added or
+  removed from the tree, _without disturbing the current layout of the tree_.
+
+
+![](20200210-design-vscode-s3-tree.png)
+
+### Create and edit S3 buckets
+
+- Root node context menu (right-click):
+  - Create bucket
+- Bucket node context menu (right-click):
+  - Delete bucket
+  - New folder...
+  - Upload...
+  - Copy name
+  - Copy ARN
+
+### Create and edit S3 objects
+
+- S3 object node context menu (right-click):
+  - Delete object
+  - Copy name
+  - Copy ARN
+  - Copy URL
+  - Upload...
+  - Edit
+    - Downloads the object to a temp folder and opens it in a new VSCode editor tab.
+  - Download
+    - Does _not_ present a chooser.
+    - Downloads immediately to the OS default "Downloads" location (typically `~/Downloads`).
+    - Download status is given as a toaster, which contains a "Choose location..." link.
+      - User can change the download location via this link. This will be the
+        default location for future downloads.
+      - The default download location is also configurable as a global Toolkit
+        option. This is not S3-specific, it is used for all "Download"
+        experiences in the Toolkit.
+
+### Drag-drop from client to server
+
+- Drag a folder from the local system to the S3 tree:
+  - onto a bucket uploads as a new top-level folder in the bucket
+  - onto a folder uploads as a new subfolder
+- Drag a file from the local system to the S3 tree:
+  - onto a bucket uploads the file as a top-level object in the bucket
+  - onto a folder uploads the file as an object in the folder
+
+### Drag-drop server-side
+
+- User can copy objects or folders from one folder to another by drag-drop.
+- User can copy objects or folders _across regions_ by dragging from one region
+  to another.
+
+### S3 bucket properties
+
+Surface popular bucket properties such as "public accessible" and "versioning
+enabled" in a clear yet uncluttered manner.
+
+- Show icon (same as AWS web console) next to bucket name indicating whether it
+  is publicly accessible
+- Show icon or sigil next to bucket name indicating whether versioning is enabled
+
+### S3 object properties
+
+S3 objects have properties like size, last-modified-date, and previous versions
```

Comment:
I'm thinking that size and last modified date are probably among the more important pieces of metadata for S3. How would you surface those? I'd personally love to see it in a tooltip attached to the object node.

Reply:
Tooltip is a "good enough" MVP for a first iteration, IMO. Although it doesn't solve the "sort" use-case.

Reply:
It seems reasonable to say that sorting is out of scope for the S3 experience based on what is being proposed so far. If a future phase introduces filtering or searching, this would assist users in locating objects of interest.